admin-dev/themes/new-theme/js/pages/product/components/categories/category-tree-selector.ts

```diff
@@ -0,0 +1,471 @@
+    this.prototypeTemplate = this.categoryTree.dataset.prototype ? this.categoryTree.dataset.prototype : null;
+    this.prototypeName = this.categoryTree.dataset.prototypeName ? this.categoryTree.dataset.prototypeName : null;
+    this.expandAllButton = this.modalContainer.querySelector(ProductCategoryMap.expandAllButton);
+    this.reduceAllButton = this.modalContainer.querySelector(ProductCategoryMap.reduceAllButton);
+    this.tags = new Tags(
+      this.eventEmitter,
+      `${ProductCategoryMap.categoriesModalContainer} ${ProductCategoryMap.tagsContainer}`,
+    );
+    this.tags.render(this.selectedCategories);
+    this.treeCategories = await getCategories();
+
+    this.initTypeaheadData(this.treeCategories, '');
+    this.initTypeahead();
+    this.initTree();
+    this.listenCancelChanges();
+    this.applyCategoryTreeChanges();
+    this.eventEmitter.on(ProductEventMap.categories.categoryRemoved, (categoryId) => this.unselectCategory(categoryId));
+  }
+
+  private applyCategoryTreeChanges(): void {
+    if (!this.modalContainer) {
+      return;
+    }
+
+    const applyBtn = this.modalContainer.querySelector(ProductCategoryMap.applyCategoriesBtn) as HTMLElement;
+
+    applyBtn.addEventListener('click', () => {
+      this.eventEmitter.emit(ProductEventMap.categories.applyCategoryTreeChanges, {
+        categories: this.selectedCategories,
+      });
+      this.closeModal();
+    });
+  }
+
+  private listenCancelChanges(): void {
+    if (!this.modalContainer) {
+      return;
+    }
+
+    const cancelBtn = this.modalContainer.querySelector(ProductCategoryMap.cancelCategoriesBtn) as HTMLElement;
+    cancelBtn.addEventListener('click', () => this.closeModal());
+  }
+
+  private initTree(): void {
+    const categoryTree = this.categoryTree;
+
+    if (!categoryTree) {
+      return;
+    }
+
+    this.treeCategories.forEach((treeCategory) => {
+      const treeCategoryElement = this.generateCategoryTree(treeCategory);
+      categoryTree.append(treeCategoryElement);
+    });
+
+    if (this.expandAllButton) {
+      this.expandAllButton.addEventListener('click', () => {
+        this.toggleAll(true);
+      });
+    }
+
+    if (this.reduceAllButton) {
+      this.reduceAllButton.addEventListener('click', () => {
+        this.toggleAll(false);
+      });
+    }
+
+    categoryTree.querySelectorAll(ProductCategoryMap.checkboxInput).forEach((checkbox) => {
+      if (checkbox instanceof HTMLInputElement) {
+        const categoryId = Number(checkbox.dataset.id);
+
+        if (this.selectedCategories.some((category) => category.id === categoryId)) {
+          checkbox.checked = true;
+        }
+
+        checkbox.addEventListener('change', (e) => {
+          const currentTarget = e.currentTarget as HTMLInputElement;
+
+          // do not allow unchecking main category id
+          //@todo: out of scope - this behavior is not ux friendly. Implement automatic closest parent selection instead (both in tree and in tags)?
```

Comment:
`So it might become bothersome to do round trip between the page and the modal just to change the default category`

yep, indeed, thats why we need to clarify the expected behavior. (I agree about the technical part)